nifpga/session.py

```diff
@@ -456,226 +449,214 @@ def __init__(self,
+        if self._overflow_enabled:
+            """ If overflow status is enabled we need an extra bit. """
+            bits_required += 1
+        if bits_required >= 32:
+            return int(ceil(bits_required / 32.0))
+        else:
+            return 1
 
     def read(self):
-        """ Reads the entire FXP number as an array U32 bits then combining
-        the array bits into a single datatype, then convert it into python
-        Decimal. The largest expecting transfer would be a 64 bit fixed point
-        with the overflow status enabled (making an array of length 3).
+        """ Reads the fixed point representation from the register
+        offset, then converts it into a more native python experience.The
+        numeric value is returned as pythons Decimal. If the register has
+        the overflow status enabled the return value is a tuple, where the
+        first element is a bool for the overflow status, and the second is the
+        decimal value.
 
         Returns:
-            data (DataType.value): The data in the register in a python list.
+            decimal.Decimal: Decimal representation of the fixed point number.
+        Returns:
+            (bool, decimal.Decimal): Boolean for overflow, Decimal
+                                     representation of the fixed point number.
         """
-        data = super(_FxpRegister, self).read()
-        data = self._combine_array_of_u32_into_one_value(data)
-        return self._convert_from_read_value_to_decimal(data)
+        buf = self._ctype_type()
+        self._read_func(self._session, self._resource, buf, self._transfer_len)
+        read_array = [elem for elem in buf]
+        fxp_integer = self._combine_array_of_u32_into_one_value(read_array)
+        return self._convert_from_read_value_to_decimal(fxp_integer)
 
     def _combine_array_of_u32_into_one_value(self, data):
+        """ This method is a helper to convert the array read from hardware
+        and return a single value removing any excessive bits. Because the
+        significant bits are left associated we need to shift the the combined
+        data to the right.
+        For example, if the register had a word length of 54, the 54 MSB would
+        be the fixed point bits. The 10 LSB of the combinedData must be shifted
+        off in order to not mess up further calculations.
+
+        """
         combinedData = 0
-        for index in range(0, len(self)):
-            combinedData = (combinedData << 32 * index) + data[index]
-        if len(self) > 1:
+        for index in range(0, self._transfer_len):
+            combinedData = (combinedData << 32) + data[index]
+        if self._transfer_len > 1:
             overflow_bit = 1 if self._overflow_enabled else 0
             total_num_bits = self._word_length + overflow_bit
-            combinedData = combinedData >> (32 * len(self) - total_num_bits)
+            combinedData = combinedData >> (32 * self._transfer_len - total_num_bits)
```

Comment:
Apparently, the data is only left-justified when _transfer_len > 1. Your comment above didn't mention that.

Reply:
Correct, Adding comment